sim/core/aura.go

```diff
@@ -386,16 +379,22 @@ func (at *auraTracker) reset(sim *Simulation) {
 }
 
 func (at *auraTracker) advance(sim *Simulation) {
-	// Loop in reverse order so that aura removal is (mostly) safe.
-	for i := len(at.activeAuras) - 1; i >= 0; i-- {
-		// Need to check again because some effects can expire 2 auras at once.
-		if i < len(at.activeAuras) {
-			aura := at.activeAuras[i]
-			if aura.expires <= sim.CurrentTime {
-				aura.Deactivate(sim)
-			}
+	if at.minExpires > sim.CurrentTime {
+		return
+	}
+
+restart:
+	minExpires := NeverExpires
+	for _, aura := range at.activeAuras {
+		if aura.expires <= sim.CurrentTime {
+			aura.Deactivate(sim)
+			goto restart // activeAuras have changed
+		}
+		if aura.expires < minExpires {
+			minExpires = aura.expires
 		}
 	}
+	at.minExpires = minExpires
 }
```

Comment:
Using "goto" since I think it's easier to understand than recursion, or adding an outer loop. Iterating in order is quite a bit faster than a reverse iteration, and in 99% of all cases (at least for the rogue sim I instrumented) nothing is deactivated, so the extra recursion(s) don't matter.

Reply:
FWIW, the case that forced me to add the extra check was Badge of the Swarmguard. The expiration of the active aura triggers the expiration of the stacks aura, so both expire at once.